### DISCUSSION TITLE
Filter

### DISCUSSION THREAD

**Discussion by michaelchambersjnr** [ministryofjustice/moj-frontend#197] [RELEASED: Components, pages and patterns]
# Existing

#### Components and patterns

- [Filter](https://design-patterns.service.justice.gov.uk/components/filter/) component in the MOJ Pattern Library
- [Filter a list](https://design-patterns.service.justice.gov.uk/patterns/filter-a-list/) pattern in the MOJ Pattern Library
- [Option select](https://finder-frontend.herokuapp.com/component-guide/option-select) in the GOV.UK Finder Frontend Component Guide

#### GitHub discussions

- [Filter a list](https://github.com/alphagov/govuk-design-system-backlog/issues/133) in the GOV.UK Design System community backlog
- [Filter a data table](https://github.com/UKHomeOffice/design-system/issues/264) issue on GitHub is a new pattern being worked on at the Home Office

#### New generic filter guidance

Do user research to find out the best order for the filters.

# Variation

### Product created in

OPG Sirius Supervision workflow

### About the product 

OPG Sirius Supervision workflow is a product being developed in OPG to replace its legacy system. 

### Who the users are

It will be used by staff of the OPG to allow them to supervise deputies appointed by the court of protection.

### User need

There is a need for users to be able to filter lists of information in order to manage their work. There are many filters and options users need to filter by. Users are unlikely to select more than one filter at a time.

### Using the existing filter component

Before designing a variation, the team first followed the guidance of the [filter component](https://design-patterns.service.justice.gov.uk/components/filter/) and [Filter a list](https://design-patterns.service.justice.gov.uk/patterns/filter-a-list/) pattern in the MOJ Pattern Library.

Research showed:

- users expected the 'Apply filters' button at the bottom. Some users thought filters would auto apply and were not expecting to have to click 'Apply filters'
- users often missed the 'Show filters' button, or assumed that the headline stats or the column headers would be the filters before realising there was a show filters option
- the component would break the page layout when a lot of filters or options in filters are needed. Task types/Case owner can be up to 20-32 options respectively.

![opg-sirius-prototypes herokuapp com_sandbox_filters (1)](https://user-images.githubusercontent.com/6122118/124463944-7e0e6f80-dd8b-11eb-9959-22935f8f012d.png)

### Designing a variation

The team used their research and the [Option select](https://finder-frontend.herokuapp.com/component-guide/option-select) component from the Finder Frontend Component Guide  as a reference to design a variation of the filter component that would meet the user need.

This variation:

- lets users show and hide each filter, similar to the Accordion component in the GOV.UK Design System
- includes a scrollable list in each filter, to allow many options to be included
- includes a filter search, to allow users to narrow the checkboxes down
- moves the 'Apply filters' button to the bottom, responding to user research

![screenshot-localhost_3000-2021 03 22-14_33_42](https://user-images.githubusercontent.com/6122118/124605717-1e37c780-de64-11eb-82b4-5cc9e7b7689a.png)

### Research on this variation

This variation has been tested with users and iterated and improved based on research.

### Is it useful and unique?

- There seems to be a need in a number of products for many filters or filter options, such as case working systems.
- It does not replicate something already in the Design System. [Filter a list](https://github.com/alphagov/govuk-design-system-backlog/issues/133) is in the GOV.UK Design System community backlog.

### When to use

- When there is a need for many filters or filter options
- When users use a service regularly, for example users of caseworking systems who need to perform familiar tasks quickly

#### Find out how your users interact with the content

Users might need the filters they have opened to stay open if they leave and then return to the page. You can configure the filters to stay open.

#### Use clear labels

Hiding filters hides the options, so the labels need to be clear. If necessary, you can add a summary line to help users understand what is in the filter.

If you struggle to come up with clear labels, it might be because the way you’ve separated the content is not clear.

# Contribute
#### Accessibility testing

Once we have a production version ready of this variation we will conduct accessibility testing.

#### Users don’t always see that there is a filter

Should we:

- show the component by default with the option to hide
- make the 'Show filters' button more prominent, for example use the blue

#### Selected filters can't be seen when filters is hidden

Should we:

- move selected filters and the ability to clear them to the top of the content area, like the [GOV.UK Search results](https://www.gov.uk/search/all?keywords=test&order=relevance)

**Comment by IxDbloke** [comment]
Hi, this pattern is ideal for a Home Office project I've been working on. I'm finishing up soon but I'll let some other designers know, along with the developers in my team.

**Comment by kelliedesigner** [comment]
Issue created to add to the Pattern Library: https://github.com/ministryofjustice/moj-frontend/issues/231

**Comment by kelliedesigner** [comment]
PR for adding to the Pattern Library: https://github.com/ministryofjustice/moj-frontend/pull/230

**Comment by kelliedesigner** [comment]
Request for showing an example of filters being used with a sortable table at the same time.

**Comment by furnivall** [comment]
Hi there,

Our service (Court and Tribunal Hearings within HMCTS) recently went through the Accessibility Audit process from the Digital Accessibility Centre (DAC).

We use the nunjucks version of the component within an alphabetical search page. We received the following feedback on the filter component:

"Users are presented with the ‘Selected filter’ section and the ‘Apply filters’ button and ‘Clear filters’ links above the filter selection checkboxes, this means that users have to navigate through the filter sections to make a selection and navigate back through the checkboxes to apply the filter selection. "

We thought it might assist other teams who decide to use this component to pre-warn them about this. It is of course possible to reorder the display of the component by manipulating the raw HTML but the default implementation is likely to receive a similar response from DAC audits.

Please let me know if I can clarify anything.

**Comment by gregtyler** [comment]
In OPG we've had to write custom JavaScript to modify this component to allow us to overrule the default open/close state of the filters. We wanted the filters to be hidden by default, and only shown when either the button is clicked or the user already has filters selected (e.g. on reload or back button) ([see PR](https://github.com/ministryofjustice/opg-sirius-lpa-frontend/pull/237)).

I think this is particularly important for apps that handle filters/pagination on the server-side, so that they can provide a consistent UI.

**Comment by Whersi** [comment]
## Ticket number

https://github.com/ministryofjustice/moj-frontend/issues/160

The details below is pulled from the ticket.



## Summary

The Filter component should provide a way of specifying custom classes.

## Motivation

Most (all?) other components offer the ability to add extra classes by providing a `classes` parameter when instantiating. This is useful for extending styling without overwriting default implementation (for instance, by providing a modifier class to create a bespoke theme).

The Filter component is currently missing this feature.

## Describe alternatives you've considered

At the moment, it seems the only way to tweak the styling of this component is to override the default styles:

#### App css:
```scss
.moj-filter__header {
  background: govuk-colour("light-grey");
}
```

When allowing extra classes, the modifier could be used to ensure original styles are not touched:

#### Template:
```
{{ mojFilter({
  heading: {
    text: 'Filter'
  },
  classes: "moj-filter--light-header",
  optionsHtml: filterOptionsHtml
}) }}
```

```scss
.moj-filter--light-header .moj-filter__header {
  background: govuk-colour("light-grey");
}
``` 

## Additional context

Happy to provide the PR for this!

**Comment by Whersi** [comment]
## Ticket number

https://github.com/ministryofjustice/moj-frontend/issues/616.

The details below is pulled from the ticket.


## Summary

Similar to [this issue](https://github.com/ministryofjustice/moj-frontend/issues/611), the Apply Filters button in the Filter component should be able to have custom classes applied to it.

## Motivation

Most other components offer the ability to add extra classes by providing a classes parameter when instantiating including the govUkButton component which is being used under the hood in the moj filter component. This is useful for extending styling without overwriting default implementation (for instance, by providing a modifier class to create a bespoke theme).

The Filter component does not allow a custom button class to be passed through to the govUkButton component used to apply the filter.

## Describe alternatives you've considered

At the moment, it seems the only way would be to tweak the styling of the gov uk button on the page it is used:

#### App css:
```scss
.govuk-button {
  custom styles
}
```

which is very undesirable as it would affect all buttons on the page and could be hidden from future developers working on the page.

Otherwise we would have to avoid using nunjucks and simply write the HTML with custom classes ourselves.

## Additional context

Happy to provide the PR for this!

**Comment by Whersi** [comment]
##  Suggest a change form - Andreea Vlad

> for the section *Navigating filters -> Show and hide individual filters* it will be useful to have the code available so that we can speed up the work. This is similar with the example box at the top that has HTML and nunjucks code. 

![Screenshot Capture - 2025-04-29 - 09-42-25](https://github.com/user-attachments/assets/11e787bc-ba65-4504-a1f1-529b2ad06fba)

**Comment by emma-nicol** [comment]
### ****Suggested change to position of apply button**** 

Other government design systems display the button below the filters not above: 
- https://design.homeoffice.gov.uk/design-system/patterns/help-users-to/search-for-something
- https://design.education.gov.uk/design-system/components/filter

**For accessibility reasons the CTA should be below the filters**** 

DFE refers to this in their design system 

_'Place the button at the bottom of the filters. This is helpful for all users.
It stops them having to navigate down to select filters, and then back up to apply them.
DfE accessibility audits have highlighted the button being at the top as an issue for people using assitive technology._

**Research within MOJ has also raised issues with the CTA being above the filters** 

'this button is located above the actual filters—this means that keyboard users would have to tab back up to the button after making their filters selection.

This would affect screen reader users, who might not understand that they need to tab back up to apply their selection. They may abandon the function thinking that it isn’t functional.

This would also be frustrating for keyboard users, since this would involve unnecessarily having to use their keyboard for additional navigation. For some users with mobility impairments, interacting with the keyboard can be a painful experience and it is important to ensure an effective and streamlined navigation process."

https://uv2854-moj-making-recall-decisions.uservisionaccessibility.co.uk/inconvenient-positioning-of-apply-filters-button-m.html

Recommendation: 

Amend the component so the CTA is below the filters see examples: 

<img width="222" height="354" alt="Screenshot 2025-10-16 at 14 12 49" src="https://github.com/user-attachments/assets/8fdad0e8-82ba-403a-9b07-fdac536177d9" />

<img width="222" height="544" alt="Screenshot 2025-10-15 at 14 01 27" src="https://github.com/user-attachments/assets/540a50cc-0a9d-4de7-b3b5-c17e028f4f49" />

**Comment by chrispymm** [comment]
DfE recently shared some work they did around filtering:

- [Improving how candidates can filter and sort courses](https://becoming-a-teacher.design-history.education.gov.uk/find-teacher-training/improving-filtering-and-sorting/)
- [Researching filters across government](https://becoming-a-teacher.design-history.education.gov.uk/find-teacher-training/researching-filters-across-government/)
- [User research with candidates on filtering and sorting courses](https://becoming-a-teacher.design-history.education.gov.uk/find-teacher-training/filtering-and-sorting-ur/)

**Comment by ethan-mills-ukef** [comment]
We've had a DAC audit on a page using this component with the following:

> The heading (Selected filters) is a subheading of the heading (Filter) however, there is no 
clear relationship between them that is reflected in the heading levels. Both headings sit at 
level two.

> Screen reader user comments:
“Once I expanded the accordion for the filters section additional headings appeared on the 
page. The two headings announced as ‘Filters’ and ‘Selected filters’ and both appeared as h2 
headings. This made the headings structure illogical as the ‘Selected filters’ heading was a 
sub-section of the ‘Filters’ heading, but this was not conveyed to me with the screen reader 
software as the heading levels were the same.
This may prevent some less experienced screen reader users from easily understanding the 
structure of the page as the heading levels do not reflect the visual structure.
Presenting the ‘Selected filters’ heading as an h3 heading will enable me to more easily 
understand the layout of the page. This issue was consistent with JAWS and NVDA.”

> Solution:
Ensure information, structure, and relationships conveyed through presentation can be 
programmatically determined and that there is a clear relationship between the h2 heading 
‘Filter’ and the subheading ‘Selected filters’ that follows it. Change the heading (Selected 
filters) to a h3

This seems pretty reasonable to me, and I'd be happy to raise a PR for this if it was likely to be accepted.

**Comment by helennickols** [comment]
In the next iteration, explore including the requirement for saving a filtered list as a custom view in their service. This is to prevent them  needing to re-filter.
